### MNIST Classification projects

In this project, I'm assigned to do classification on the MNIST handwritten digit, this is an old dataset for any Machine Learning Beginner. 

At the moment, I've learnt KNN, Naive Bayes, Decision Tree and SVM (convert to dual problem and solve by quadratic programming). After the first try with SVM quadratic programming with the following objective function:

$$ g(\lambda, \mu) = \min_{\mathbf{w}, b, \xi} \mathcal{L}(\mathbf{w}, b, \xi, \lambda, \mu) $$
$$ g(\lambda, \mu) = \sum_{n=1}^N \lambda_n - \frac{1}{2} \sum_{n=1}^N\sum_{m=1}^N \lambda_n \lambda_m y_n y_m k(\mathbf{x}_n^T, \mathbf{x}_m) $$

\begin{eqnarray}
     \lambda &=& \arg \max_{\lambda} g(\lambda)   &&\\
     \text{subject to:}~ && \sum_{n=1}^N \lambda_ny_n = 0 && \\
     && 0 \leq \lambda_n \leq C, ~\forall n= 1, 2, \dots, N && 
\end{eqnarray}

I realize that it's impossible to compute $ k(\mathbf{x}_n^T, \mathbf{x}_m) $ for the number of dataset = 60000. One way to overcome this difficulty is that approach the problem in another way in SVM that we can use gradient descent or iterative methods and change the loss function, I will try this approach later.

In this project, I do 2 algorithms in deep learning way: 
- Softmax Regression (no hidden layer)
- Lenet (5 hiddens layer)

--------------
 

#### 1. Load MNIST dataset

In [1]:
from utils import load_dataset_mnist, preprocess_data
from mnist import MNIST
from sklearn.preprocessing import StandardScaler

load_dataset_mnist()

mndata = MNIST('data_mnist')

images, labels = mndata.load_training()

-------> Downloading MNIST dataset
-------> Finish


#### 2. Training or Loading Lenet Model

In [2]:
import os
from lenet import Lenet
import tensorflow as tf

training_phase = "saved_model" not in os.listdir()

lenet = Lenet(20, 64, tf.train.AdamOptimizer(learning_rate=0.001), tf.losses.softmax_cross_entropy)

if training_phase:
    images, labels = mndata.load_training()
    images, labels = preprocess_data(images, labels, True)
    lenet.train(images, labels)
else:
    images_test, labels_test = mndata.load_testing()
    images_test, labels_test = preprocess_data(images_test, labels_test, True, True)

    pred = lenet.predict(images_test)
    print("Accuracy:", len(labels_test[pred == labels_test]) / len(labels_test))  # 98%

    from sklearn.metrics.classification import confusion_matrix
    print("Confusion matrix: ")
    print(confusion_matrix(labels_test, pred))


Instructions for updating:
Colocations handled automatically by placer.
Accuracy: 0.983
Confusion matrix: 
[[ 970    0    2    0    0    2    1    1    1    3]
 [   0 1127    1    2    1    1    1    0    2    0]
 [   7    2 1010    2    2    1    2    4    2    0]
 [   0    0    4  995    0    2    0    2    5    2]
 [   1    0    2    0  971    0    3    0    0    5]
 [   1    0    0    3    0  883    4    1    0    0]
 [   4    3    1    0    1    6  939    0    4    0]
 [   0    4   13    2    0    1    0 1001    2    5]
 [   3    0    2    3    2    4    0    3  953    4]
 [   0    3    0    4    9    6    0    5    1  981]]


------------------
Accuracy is 98% with 20 epochs. Very well!

#### 3. Training with Softmax Regression

In [3]:
from softmax_regression import SoftmaxRegression

images, labels = mndata.load_training()
images, labels = preprocess_data(images, labels)
softmax = SoftmaxRegression(epochs=20)
softmax.train(images, labels)

images_test, labels_test = mndata.load_testing()
images_test, labels_test = preprocess_data(images_test, labels_test, test=True)

pred = softmax.predict(images_test)

print("Accuracy:", len(pred[labels_test == pred]) / len(pred))
from sklearn.metrics.classification import confusion_matrix
print("Confusion matrix: ")
print(confusion_matrix(labels_test, pred))

/home/tgiang/Desktop/framgia-training/venv/lib/python3.5/site-packages/sklearn/preprocessing/_encoders.py:371: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
  warnings.warn(msg, FutureWarning)


Loss at epoch 1 14.57
Loss at epoch 2 6.11
Loss at epoch 3 4.53
Loss at epoch 4 3.77
Loss at epoch 5 3.32
Loss at epoch 6 3.01
Loss at epoch 7 2.77
Loss at epoch 8 2.58
Loss at epoch 9 2.44
Loss at epoch 10 2.31
Loss at epoch 11 2.20
Loss at epoch 12 2.11
Loss at epoch 13 2.02
Loss at epoch 14 1.95
Loss at epoch 15 1.88
Loss at epoch 16 1.85
Loss at epoch 17 1.79
Loss at epoch 18 1.74
Loss at epoch 19 1.69
Loss at epoch 20 1.65
Accuracy: 0.8721
Confusion matrix: 
[[ 920    0    3    6    1   19   20    5    6    0]
 [   0 1073   12    6    1    5    5    8   24    1]
 [  16   18  877   29   21    6   19   14   29    3]
 [   5   10   27  860    3   52    6   16   20   11]
 [   2    4    6    3  874    3   15   15    7   53]
 [  16    5    3   38   13  727   25   17   39    9]
 [  16    4   13    1   15   23  878    4    3    1]
 [   3   14   26   14    5    4    3  896    4   59]
 [  16   20   16   30   32   35   16   17  779   13]
 [  13    7    3   15   64   13    1   44   12  837]]


With softmax regression the accuracy is 87%! Not so bad.